pubkey/src/lib.rs

```diff
@@ -158,10 +160,133 @@ impl From<u64> for PubkeyError {
 #[cfg_attr(all(feature = "borsh", feature = "std"), derive(BorshSchema))]
 #[cfg_attr(feature = "serde", derive(Deserialize, Serialize))]
 #[cfg_attr(feature = "bytemuck", derive(Pod, Zeroable))]
-#[derive(Clone, Copy, Default, Eq, Hash, Ord, PartialEq, PartialOrd)]
+#[derive(Clone, Copy, Default, Eq, Ord, PartialEq, PartialOrd)]
 #[cfg_attr(feature = "dev-context-only-utils", derive(Arbitrary))]
 pub struct Pubkey(pub(crate) [u8; 32]);
 
+/// Custom impl of Hash for Pubkey
+/// allows us to skip hashing the length of the pubkey
+/// which is always the same anyway
+impl Hash for Pubkey {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        state.write(self.as_array());
+    }
+}
+
+#[cfg(all(feature = "rand", not(target_os = "solana")))]
+mod hasher {
+    use rand::Rng;
+
+    /// A special Hasher specifically for Pubkeys. Uses random 8 bytes as
+    /// input. Since pubkeys are effectively random, hashing them is pure overhead.
+    /// Use with care, this may be abusable.
```

Comment:
The documentation contradicts itself. Hashing is pure overhead, but not hashing can be abused?

I'd say something like 

```
A faster, but less collision resistant hasher for pubkeys.

Specialized hasher that uses a random 8 bytes subslice of the pubkey as the hash value. Should not be used when collisions might be used to mount DOS attacks.
```

Reply:
applied in https://github.com/anza-xyz/solana-sdk/pull/96/commits/241d97a2bea2a2fa07efcaf23ccd555a8dc5f06c